File: app/src/main/java/bogomolov/aa/fitrack/model/RamerDouglasPeucker.java
package bogomolov.aa.fitrack.model;

import android.util.Log;

import java.util.ArrayList;
import java.util.List;

public class RamerDouglasPeucker {


    private static double perpendicularDistance(double px, double py, double vx, double vy, double wx, double wy) {
        final double l2 = (vx - wx) * (vx - wx) + (vy - wy) * (vy - wy);
        if (l2 == 0)
            return Point.distance(new Point(px, py), new Point(vx, vy));
        final double t = ((px - vx) * (wx - vx) + (py - vy) * (wy - vy)) / l2;
        if (t < 0)
            return Point.distance(new Point(px, py), new Point(vx, vy));
        if (t > 1)
            return Point.distance(new Point(px, py), new Point(wx, wy));
        return Point.distance(new Point(px, py), new Point((vx + t * (wx - vx)), (vy + t * (wy - vy))));
    }

    private static void douglasPeucker(List<Point> list, int s, int e, double epsilon, List<Point> resultList) {
        double dmax = 0;
        int index = 0;

        final int start = s;
        final int end = e - 1;
        for (int i = start + 1; i < end; i++) {
            Point p = list.get(i);
            Point v = list.get(start);
            Point w = list.get(end);
            final double d = perpendicularDistance(p.getLat(), p.getLng(), v.getLat(), v.getLng(), w.getLat(), w.getLng());
            if (d > dmax) {
                index = i;
                dmax = d;
            }
        }
        //if (e - s < 5) TrackerService.stringBuffer.append("perpendicularDistance " + dmax + " epsilon " + epsilon + "\n");
        if (dmax > epsilon) {
            douglasPeucker(list, s, index, epsilon, resultList);
            douglasPeucker(list, index, e, epsilon, resultList);
        } else {
            if ((end - start) > 0) {
                resultList.add(list.get(start));
                resultList.add(list.get(end));
            } else {
                resultList.add(list.get(start));
            }
        }
    }

    public static List<Point> douglasPeucker(List<Point> list, double epsilon) {
        if (list.size() < 3) return list;
        List<Point> resultList = new ArrayList<>();
        douglasPeucker(list, 0, list.size(), epsilon, resultList);
        return resultList;
    }
}

File: app/src/main/java/bogomolov/aa/fitrack/core/RamerDouglasPeucker.kt
package bogomolov.aa.fitrack.core

import bogomolov.aa.fitrack.core.model.Point
import java.util.ArrayList

private fun perpendicularDistance(px: Double, py: Double, vx: Double, vy: Double, wx: Double, wy: Double): Double {
    val l2 = (vx - wx) * (vx - wx) + (vy - wy) * (vy - wy)
    if (l2 == 0.0)
        return Point.distance(Point(px, py), Point(vx, vy))
    val t = ((px - vx) * (wx - vx) + (py - vy) * (wy - vy)) / l2
    if (t < 0)
        return Point.distance(Point(px, py), Point(vx, vy))
    return if (t > 1) Point.distance(Point(px, py), Point(wx, wy)) else Point.distance(Point(px, py), Point(vx + t * (wx - vx), vy + t * (wy - vy)))
}

private fun douglasPeucker(list: List<Point>, s: Int, e: Int, epsilon: Double, resultList: MutableList<Point>){
    var dmax = 0.0;
    var index = 0;

    val start = s;
    val end = e-1;
    for(i in start+1 until end){
        val p = list[i]
        val v = list[start]
        val w = list[end]
        val d = perpendicularDistance(p.lat, p.lng, v.lat, v.lng, w.lat, w.lng)
        if(d > dmax){
            index = i
            dmax = d
        }
        if(dmax > epsilon){
            douglasPeucker(list, s, index, epsilon, resultList)
            douglasPeucker(list, index, e, epsilon, resultList)
        }else{
            resultList.add(list[start])
        }
    }
}

fun douglasPeucker(list: List<Point>, epsilon: Double): List<Point> {
    if (list.size < 3) return list
    val resultList = ArrayList<Point>()
    douglasPeucker(list, 0, list.size, epsilon, resultList)
    return resultList
}